### ISSUE TITLE
Add a parameterized AMP argument type which can serialize and unserialize 0 or more of a specified type

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4116]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4116|
|Type|enhancement|
|Created|2009-11-17 21:43:26Z|
|Branch|https://github.com/twisted/twisted/tree/amp-listof-4116|

There's presently no argument type which supports several integers, or several strings.  Strangely, there is an argument type for several dictionaries (`AmpList`).  Supporting some other types would be nice.

This might look something like `ListOf(String())`.  `AmpList` might be expressed in terms of this as `ListOf(Dict())`, although there is presently no `Dict` argument type.

I'm most immediately motivated to have such an argument type for [#1784](https://github.com/twisted/twisted/issues/1784): code for that ticket smashes a list of strings together in one process and then splits it up again in another process.  Supremely gross.  I'm certain there are many other similar use cases, though.

<details><summary>Searchable metadata</summary>

```
trac-id__4116 4116
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_amp_listof_4116 branches/amp-listof-4116
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp amp
time__1258494206000000 1258494206000000
changetime__1259904479000000 1259904479000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27533]) Branching to 'amp-listof-4116'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [27534]) Add `ListOf.toBox` and `ListOf.fromBox` tests

refs [#4116](https://github.com/twisted/twisted/issues/4116)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [27537]) Fix a few small test mistakes, adjust the expected interface of ListOf to only accept Argument instances, and implement it

refs [#4116](https://github.com/twisted/twisted/issues/4116)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

 * ```
+    The serialized form of the entire list is subject to the limit imposed by
+    L{MAX_VALUE_LIMIT}.  List elements are represented as 16-bit length
```

The variable is actually called `MAX_VALUE_LENGTH`. And I think this limit is not actually enforced, as you're not calling serialize but packing directly.

 * It would be nice to have a test for another type of objects (not just strings).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

Thanks!

I fixed the docstring.  I think the limit will be enforced, just not by any code I added.  Eventually the string returned by `toString` will end up in an `AmpBox` and that will get serialized, and there's a length check there.  This is all I was trying to convey.  Do you think it's confusing to mention that limit here?

I also added tests for `Integer` and `Unicode`.

All that in r27542.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix set owner to @exarkun|
|-|-|

1 You should be more explicit about the fact that this only works with Argument subclasses which only implement fromString and toString and don't do anything weird with the box.

2 It kinda sucks that you're calling pack("!H") instead of using some other code that calls that for you.


Just fix the first problem and then you can merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [27671]) Be explicit about what kind of argument type objects are going to work with ListOf

refs [#4116](https://github.com/twisted/twisted/issues/4116)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [27676]) Merge amp-listof-4116

Author: exarkun
Reviewer: therve, radix
Fixes: [#4116](https://github.com/twisted/twisted/issues/4116)

Add a new `ListOf` AMP argument type which allows the creation of new
sequence argument types by composition with other AMP argument types.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
